emp-churn-step-by-step/src/app13.py

```diff
@@ -0,0 +1,390 @@
+# Step 13
+# Add tabs and tabs switching
+# ---
+from h2o_wave import main, app, Q, ui, on, handle_on, data
+from h2o_wave.core import expando_to_dict
+
+import altair
+import numpy as np
+import pandas as pd
+
+import logging
+from os import environ
+import traceback
+
+log = logging.getLogger("app")
+rows_per_page = 10
+
+
+def setup_logger():
+    """
+    Default Logging Level is INFO. To overwrite set environment variable LOG_LEVEL
+    Available options: INFO, DEBUG, ....
+    :return:
+    """
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.INFO)
+    formatter = logging.Formatter(
+        "%(asctime)s - [%(name)s] - %(levelname)s - %(message)s"
+    )
+    ch.setFormatter(formatter)
+    log.addHandler(ch)
+    log.setLevel(environ.get("LOG_LEVEL", "INFO").upper())
+
+
+def on_startup():
+    """
+    Initializer logger
+    :return:
+    """
+    setup_logger()
+    log.info('+++++++App started!++++++++++')
+
+
+@app('/', on_startup=on_startup, mode='unicast')
+async def serve(q: Q):
+    try:
+        # log.info - always printed
+        # log.debug - printed only when LOG_LEVEL is set to DEBUG
+        log.info("====== Start serve Function ========")
+        q_args_dict = expando_to_dict(q.args)
+        log.info(">>>> q.args >>>>")
+        for k, v in q_args_dict.items():
+            log.info(f"{k}: {v}")
+        log.info("<<<< q.args <<<<")
+        log.info(f"q.user: {q.user}")
+        log.info(f"q.events: {q.events}")
+
+        # First time a browser(tab) opens the app
+        action_taken = False
+        if not q.client.initialized:
+            action_taken = True
+            await init(q)
+            q.client.initialized = True
+            await q.page.save()
+        elif q.args['#'] is not None and q.args['#'] != q.client.current_tab:
+            log.info("========Handle  Tab Change=====")
+            log.info(f"Previous tab saved in q.client.current_tab: {q.client.current_tab}")
+            action_taken = True
+            # Set new active menu item
+            q.page['nav_menu'].value = f'#{q.args["#"]}'
+            q.client.current_tab = f'{q.args["#"]}'
+            if q.args['#'] == "details":
+                log.info("========Handle  Tab Change Switch to details=====")
+                # Restore threshold value to the last clicked on
+                log.info(f"previous threshold:{q.client.threshold}")
+                q.page['threshold_card'].items[0].slider.value = q.client.threshold
+                # Display a different tab
+                q.page["meta"] = q.client.details
```

Comment:
Not sure I understand how to perform page switch correctly. Aren't `q.client.cards.remove(name)` deleting the card? Please let me where I can see example of tab/page switch without re-creating the cards.

Reply:
> Aren't q.client.cards.remove(name) deleting the card

Yes, they are.

> Please let me where I can see example of tab/page switch without re-creating the cards.

The cards need to be destroyed when navigating to a new "app page". Btw, they are recreated even with this approach since the layout changed. Updating cards is good when you have a component in hindsight all the time and want to only change parts of it while preventing component flicker, losing focus etc.